app/services/add_attachment_service.rb

```diff
@@ -0,0 +1,63 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddAttachmentService
+  attr_reader :container, :author
+
+  def initialize(container, author:)
+    @container = container
+    @author = author
+  end
+
+  ##
+  # Adds and saves the uploaded file as attachment of the given container.
+  # In case the container supports it, a journal will be written.
+  #
+  # An ActiveRecord::RecordInvalid error is raised if any record can't be saved.
+  def add_attachment(uploaded_file:, description:)
+    attachment = Attachment.new(file: uploaded_file,
+                                container: container,
+                                description: description,
+                                author: author)
+    save attachment
```

Comment:
Lets just get one thing out of the way, to avoid a misunderstanding:

> In the current state, you get the attachment from the service, even though you don't need it

1. I only get the attachment on the happy path (not in the error case, which raises an exception)
2. I actually need the attachment, because I want to know the attachment I just added (and the caller is using it)

On to the general talk:

I am personally a fan of returning things in the successful case and raising exceptions in the unsuccessful case. Whats also okay is a method that returns if it was successful or failed, but what is (IMO) inapropriate is a method that returns a thing (e.g. an attachment) in the successful case and false in the failing case (PHP-esque return style).
We can discuss if there should be a specifically crafted exception for the error case. My thought was that the `ActiveRecord::RecordInvalid` exception would be well known and already fitting the purpose very well. Therefore I documented it as expected exception for this method. I usually only come up with own exceptions if the framework does:

* not provide a fitting exception
* I want the caller to save from knowing which kinds of exception there are to be expected ([something I earned some flak for on SO ^^](http://stackoverflow.com/questions/27185666/exception-messages-for-exception-nesting))

Another thing in your proposal that I wanted to avoid is *state*. Or at least I used different state than you are proposing ^^

In my thought you would create the service for a user and a WP and then could add one or more attachments via `add_attachment`. Each call to `add_attachment` would be self-contained.

Your proposal ditches the "reusability" of the service and creates are very specific service for that one attachment. The upside being that I can now `call` the service like I would call a block or lambda. However in your proposal the errors that occured are stored in the service and need to be picked up after `call`. This is something I am not a big fan of. Reminds me of [GetLastError](https://msdn.microsoft.com/de-de/library/windows/desktop/ms679360(v=vs.85).aspx) ^^